### PR TITLE
fix for subscribing to topics with wildcards

### PR DESCRIPTION
**Pull request by dubaxx**
retains old hashmap logic.

### CODE REVIEW

**Comment by dubaxx** [comment]
Just realized I made it a bit overcomplicated with the splitting and adding/removing slashes. Not sure what I was thinking there.

File: src/main/java/com/clearblade/java/api/MqttClient.java [hunk 1]
```diff
@@ -321,11 +322,32 @@ public void connectionLost(Throwable arg0) {
...
+				List<String> matching = keys.stream()
+						.filter(pattern.asPredicate())
+						.collect(Collectors.toList());//check key set against keys with regex
+				allMatches.addAll(matching);
+			});
+			if (allMatches.size() == 1) {//if there's more than one match, there's a problem
+				callback = callbackByTopic.get(allMatches.get(0));
+				callback.done(topic, message.getPayload());
+				callback.done(topic, new String(message.getPayload()));
+			}
```
Comment (danielrs) [inline src/main/java/com/clearblade/java/api/MqttClient.java]: Are we sure this code works? I added a test but it seems to fail:

```
@Test
void messageArrivedOnSubscribedWildcardTopicUsesCallback() throws Exception {
    MessageCallback mockCallback = mock(MessageCallback.class);

    spyClient.subscribe("topic-0", 0, mockCallback);
    spyClient.subscribe("level/+/topic", 0, mockCallback);
    spyClient.subscribe("multi/#", 0, mockCallback);

    MqttMessage mockMessage = mock(MqttMessage.class);

    when(mockMessage.getPayload()).thenReturn("foo".getBytes(), "bar".getBytes(), "baz".getBytes());

    spyClient.messageArrived("topic-0", mockMessage);
    spyClient.messageArrived("level/foo/topic", mockMessage);
    spyClient.messageArrived("multi/foo/topic", mockMessage);

    verify(mockCallback, times(1)).done("topic-0", "foo".getBytes());
    verify(mockCallback, times(1)).done("topic-0", "foo");

    verify(mockCallback, times(1)).done("level/foo/topic", "bar".getBytes());
    verify(mockCallback, times(1)).done("level/foo/topic", "bar");

    verify(mockCallback, times(1)).done("multi/foo/topic", "baz".getBytes());
    verify(mockCallback, times(1)).done("multi/foo/topic", "baz");
}
```

The failure error is:

```
Wanted but not invoked:
messageCallback.done(
    "level/foo/topic",
    [(byte) 0x62, (byte) 0x61, (byte) 0x72]
);
-> at com.clearblade.java.api.MqttClientTests.messageArrivedOnSubscribedWildcardTopicUsesCallback(MqttClientTests.java:173)

However, there were exactly 2 interactions with this mock:
messageCallback.done(
    "topic-0",
    [(byte) 0x66, (byte) 0x6F, (byte) 0x6F]
);
-> at com.clearblade.java.api.MqttClient.messageArrived(MqttClient.java:329)

messageCallback.done("topic-0", "foo");
-> at com.clearblade.java.api.MqttClient.messageArrived(MqttClient.java:330)


Wanted but not invoked:
messageCallback.done(
    "level/foo/topic",
    [(byte) 0x62, (byte) 0x61, (byte) 0x72]
);
-> at com.clearblade.java.api.MqttClientTests.messageArrivedOnSubscribedWildcardTopicUsesCallback(MqttClientTests.java:173)

However, there were exactly 2 interactions with this mock:
messageCallback.done(
    "topic-0",
    [(byte) 0x66, (byte) 0x6F, (byte) 0x6F]
);
-> at com.clearblade.java.api.MqttClient.messageArrived(MqttClient.java:329)

messageCallback.done("topic-0", "foo");
-> at com.clearblade.java.api.MqttClient.messageArrived(MqttClient.java:330)


	at com.clearblade.java.api.MqttClientTests.messageArrivedOnSubscribedWildcardTopicUsesCallback(MqttClientTests.java:173)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
```

**Review by danielrs — CHANGES_REQUESTED** [review]
I commited a test to make sure this feature works.

Please check `MqttClientTests` under `src/test`.

Tested using IntelliJ IDEA 20.2.3 with Java SDK 11.

File: src/main/java/com/clearblade/java/api/MqttClient.java (hunk 1 above)
Comment (dubaxx) [reply src/main/java/com/clearblade/java/api/MqttClient.java]: It was working for me when testing with test topics on the platform. I'll take a closer look at this tomorrow

**Comment by dubaxx** [comment]
@danielrs try again. Had to modify your test because for some reason multiple thenReturns wasn't playing nice. But I think multiple mock messages still correctly tests the logic.

**Review by danielrs — APPROVED** [review]
Looks good!

**Pull request merged**